### PR TITLE
Feature/d3 asim 962

### PR DESCRIPTION
**Pull request by hannesdiedrich**
This PR contains:
- calculating device statistics and storing them in a json file in the aggregates_results folder
- prpopsal for bugfix of the missing last market rotation
- one fix of a integrationtest `D3ASIM-662, IAA should always track offers back to its source`

### CODE REVIEW

File: src/d3a/d3a_core/sim_results/area_statistics.py [hunk 1]
```diff
@@ -62,7 +62,7 @@ def gather_area_loads_and_trade_prices(area, load_price_lists):
     return load_price_lists
 
 
-def gather_prices_pv_stor_energ(area, price_energ_lists):
+def gather_prices_pv_storage_energy(area, price_energ_lists):
```
Comment (spyrostz) [inline src/d3a/d3a_core/sim_results/area_statistics.py]: Very good, I did not like the truncated name as well.

File: src/d3a/models/strategy/load_hours.py [hunk 2]
```diff
@@ -112,6 +113,8 @@ def event_activate(self):
                             for day in range(self.area.config.sim_duration.days + 1)}
         self._simulation_start_timestamp = self.area.now
         self.assign_energy_requirement(self.avg_power_W)
+        self.energy_requirement_history_kWh = dict((k, v / 1000.)
```
Comment (spyrostz) [inline src/d3a/models/strategy/load_hours.py]: Is this different from self.state.desired_energy_Wh (member of LoadState class) ? If not, then it is better to reuse the member from the state class, in order to not add another class member.

File: src/d3a/d3a_core/sim_results/endpoint_buffer.py [hunk 3]
```diff
@@ -70,10 +75,17 @@ def generate_json_report(self):
             "cumulative_grid_trades": self.cumulative_grid_trades,
             "bills": self.bills,
             "tree_summary": self.tree_summary,
-            "status": self.status
+            "status": self.status,
+            "device_stats": self.device_statistics_time_str_dict
         }
 
     def update_stats(self, area, simulation_status):
+        # if not (not hasattr(area.parent, "past_markets") or len(area.parent.past_markets) == 0):
```
Comment (spyrostz) [inline src/d3a/d3a_core/sim_results/endpoint_buffer.py]: Trivial, as a reminder please remove these before you merge.

File: src/d3a/d3a_core/sim_results/device_statistics.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            else:
+                self.gather_device_statistics(child, subdict[child.name])
+
+    def _gather_device_statistics(self, area: Area, subdict: Dict):
+        if not hasattr(area.parent, "past_markets") or len(area.parent.past_markets) == 0:
+            return None
+
+        if isinstance(area.strategy, PVStrategy):
+
+            self._device_price_stats(area, subdict)
```
Comment (spyrostz) [inline src/d3a/d3a_core/sim_results/device_statistics.py]: _device_price_stats and _device_energy_stats can be moved out of the if-elif blocks to remove duplication. You can check if area.strategy is not None, then you can safely deduce that this area is a device, therefore you can call safely these methods.

File: src/d3a/models/strategy/load_hours.py (hunk 2 above)
Comment (hannesdiedrich) [reply src/d3a/models/strategy/load_hours.py]: Yes, you are right, can not remember the reason for adding this here. Will try to remove it.

File: src/d3a/d3a_core/export.py [hunk 5]
```diff
@@ -93,6 +93,9 @@ def export_json_data(self, directory: dir):
         trade_file = os.path.join(json_dir, "trade-detail")
         with open(trade_file, 'w') as outfile:
             json.dump(self.endpoint_buffer.trade_details, outfile, indent=2)
+        device_file = os.path.join(json_dir, 'device_statistics')
```
Comment (faizan2590) [inline src/d3a/d3a_core/export.py]: I guess these three line could be removed as https://github.com/gridsingularity/d3a/pull/378/files#diff-9e95bed0f62db1b1f967c65f0823d49dR100 it is taking care of dumping device statistics as json

Comment (hannesdiedrich) [reply src/d3a/d3a_core/export.py]: You are right! Thanks!

**Pull request merged**